### PR TITLE
SNOW-1460206 Introduce structured arrays

### PR DESCRIPTION
**Pull request by sfc-gh-pfus**
### Description

SNOW-1460206 Introduced structrued arrays binding

### Checklist
- [x] Created tests which fail without the change (if possible)
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

File: converter.go [hunk 1]
```diff
@@ -229,52 +240,180 @@ func valueToString(v driver.Value, tsmode snowflakeType, params map[string]*stri
...
+		}
+		if len(barr) == 0 {
+			res := "[]"
+			return bindingValue{&res, "", &schemaForBytes}, nil
+		}
+		res := "["
+		for _, b := range barr {
+			res += fmt.Sprint(b) + ","
+		}
+		res = res[0:len(res)-1] + "]"
```
[the remark this answers is not in the document]
Comment (sfc-gh-pfus) [reply converter.go:367]: What exactly would you like to have extracted? We still need to test it e2e, so unit test is not enough.

File: converter.go [hunk 2]
```diff
@@ -229,52 +240,180 @@ func valueToString(v driver.Value, tsmode snowflakeType, params map[string]*stri
...
+					{
+						Type:     typ,
+						Nullable: true,
+					},
+				},
+			}}, err
+		}
+		resString := string(res)
+		return bindingValue{&resString, "json", nil}, nil
+	} else if reflect.TypeOf(v).Elem().Kind() == reflect.Struct || (reflect.TypeOf(v).Elem().Kind() == reflect.Pointer && reflect.TypeOf(v).Elem().Elem().Kind() == reflect.Struct) {
```
[the remark this answers is not in the document]
Comment (sfc-gh-pfus) [reply converter.go:314]: To be precise - array of structs or pointers to structs. I introduced a function.

File: converter.go [hunk 3]
```diff
@@ -229,52 +240,180 @@ func valueToString(v driver.Value, tsmode snowflakeType, params map[string]*stri
...
+		s := strconv.FormatBool(typedVal.Bool)
+		return bindingValue{&s, "", nil}, nil
+	case sql.NullInt64:
+		if !typedVal.Valid {
+			return bindingValue{nil, "", nil}, nil
+		}
+		s := strconv.FormatInt(typedVal.Int64, 10)
+		return bindingValue{&s, "", nil}, nil
+	case sql.NullFloat64:
+		if !typedVal.Valid {
```
[the remark this answers is not in the document]
Comment (sfc-gh-pfus) [reply converter.go:407]: I don't think so. There is no common interface or something. Go rocks :P

File: structured_type.go [hunk 4]
```diff
@@ -538,6 +538,9 @@ func (sowc *structuredObjectWriterContext) toFields() []fieldMetadata {
 type ArrayOfScanners[T sql.Scanner] []T
 
 func (st *ArrayOfScanners[T]) Scan(val any) error {
+	if val == nil {
+		return nil
```
[the remark this answers is not in the document]
Comment (sfc-gh-pfus) [reply structured_type.go:542]: Yes, this is a case when there was a null in a table.

**Review by sfc-gh-dheyman-1 — APPROVED** [review]

File: converter.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sfc-gh-pfus) [reply converter.go:367]: We decided to leave it as is.

**Pull request merged**